xonsh: Add wrapper

adisbladis opened the pull request:
###### Description of changes

It is currently not obvious how to install/use xonsh with dependencies and python packages.

This PR implements a wrapper that allows you to construct a custom xonsh environment by using:
``` nix
xonsh.override { extraPackages = ps: [ ps.requests ]; }
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented on nixos/modules/programs/xonsh.nix:
```diff
@@ -28,7 +28,7 @@ in
         default = pkgs.xonsh;
         defaultText = literalExpression "pkgs.xonsh";
-        example = literalExpression "pkgs.xonsh.override { configFile = \"/path/to/xonshrc\"; }";
+        example = literalExpression "pkgs.xonsh.override { extraPackages = ps: [ ps.requests ]; }";
```
```suggestion
        example = literalExpression "pkgs.xonsh.override { extraPackages = ps: with ps; [ requests ]; }";
```

SuperSandro2000 commented on pkgs/shells/xonsh/wrapper.nix:
```diff
@@ -0,0 +1,24 @@
+  inherit (xonsh) pname version meta;
+} ''
+  mkdir -p $out/bin
+  for bin in ${lib.getBin xonsh}/bin/*; do
+    ln -s ${pythonEnv}/bin/$(basename "$bin") $out/bin/
+  done
```
```suggestion
  lndir ${lib.getBin xonsh}/bin $out/bin
```
lndir is under xorg.lndir but not important just an idea

adisbladis replied:
If you look a bit closer you can see that these are not the same.
What I do in this wrapper is to link a subset of binaries (those available in the xonsh package) from a python environment.
This is a filtering mechanism.

SamLukeYes commented on pkgs/shells/xonsh/default.nix:
```diff
@@ -90,5 +77,6 @@ python3Packages.buildPythonApplication rec {
 
   passthru = {
     shellPath = "/bin/xonsh";
```
My NixOS configuration (https://github.com/SamLukeYes/nixos/commit/7bf704df6ea851ff68a3ad005f24178ee84e8e64) fails to build with this change.

```
error: /nix/store/0bq90g80r0b1s9s8ipznsh1azhsl8swp-xonsh-0.14.0 is not a shell package
```

Perhaps the wrapper package should pass through `shellPath`, I suppose.

adisbladis replied:
It does inherit `passthru` so I'm not sure why you're seeing this.

That error comes from `nixos/lib/utils.nix`. I copy pasted from there to make this minimal case:
``` nix
let
  pkgs = import ./. { };
  inherit (pkgs) lib;
  inherit (lib) types;

  toShellPath = shell:
    if types.shellPackage.check shell then
      "/run/current-system/sw${shell.shellPath}"
    else if types.package.check shell then
      throw "${shell} is not a shell package"
    else
      shell;

in toShellPath pkgs.xonsh

```
which works.

SamLukeYes replied:
> It does inherit `passthru` so I'm not sure why you're seeing this.

No, it doesn't.

```
nix-repl> legacyPackages.x86_64-linux.xonsh.passthru
{ }
```

> That error comes from `nixos/lib/utils.nix`. I copy pasted from there to make this minimal case:
> 
> ```nix
> let
>   pkgs = import ./. { };
>   inherit (pkgs) lib;
>   inherit (lib) types;
> 
>   toShellPath = shell:
>     if types.shellPackage.check shell then
>       "/run/current-system/sw${shell.shellPath}"
>     else if types.package.check shell then
>       throw "${shell} is not a shell package"
>     else
>       shell;
> 
> in toShellPath pkgs.xonsh
> ```
> 
> which works.

I also tried this on your branch. It throws the error.

adisbladis replied:
I forgot to push the change that I had made locally! Doh!

adisbladis committed:
xonsh: Add wrapper

It is currently not obvious how to install/use xonsh with dependencies and python packages.

This PR implements a wrapper that allows you to construct a custom xonsh environment by using:
``` nix
xonsh.override { extraPackages = ps: [ ps.requests ]; }
```

SamLukeYes reviewed (approved):
Works for me now. Thanks!
